### ISSUE TITLE
Provide way for unit testing frameworks to catch StackOverflow exception

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/runtime#40302]
## Background and Motivation

It's impossible to catch a StackOverflowException. This means that when a unit test throws one the unit testing framework just crashes. No other tests run, and it's very difficult to find out why the process crashed and on which tests - you need to go through the logs to work it out, and sometimes they don't even provide that information.

Stack Overflows are pretty common in unit tests, especially whilst in the process of writing recursive algorithms, and so it would be convenient if there was some way for a unit testing framework to say: "I know the risk, but let me catch this stack overflow exception anyway".

**Comment by Gnbrkm41** [comment]
Related: (Possibly duplicate of) https://github.com/dotnet/runtime/issues/4113 https://github.com/dotnet/runtime/issues/9183

I think providing the capability could be nice, but the problem I see is basically what @jkotas mentioned in the other issue https://github.com/dotnet/runtime/issues/9183#issuecomment-339204775 - That, there might not be enough stack space left for the handler to run.

IMO, a better approach would be to run the tests in an isolated environment of some sort (e.g. other process) and let them talk via IPC, instead of running the test on the process that the testing framework itself runs on. This way, if the process crashes, the testing framework can notice and report the problem instead of crashing with the tests. Not only that this makes it resilient against `StackOverflowException`s, this also benefits certain other scenarios such as `AccessViolationException`/`ExecutionEngineException` which cannot be caught just like `StackOverflowException` (and is much more harder situation to handle).

**Comment by GrabYourPitchforks** [comment]
I believe it would contort the runtime code if we were to attempt to make these exceptions catchable.

What if we could instead invest in making the error message more diagnosable? I think StackOverfowException now always includes the relevant stack trace, but if there are scenarios where it's not present we could try to make those a bit better. Same with AccessViolationException: maybe we could report the full stack trace along with the memory address that triggered the AV.

**Comment by Gnbrkm41** [comment]
I think the logs do have enough info (e.g. stack trace for both AVE/Stack Overflow) for diagnosis, it's just that certain toolings don't expect things to just crash without reporting and is unintuitive.

For example, if the test process crashes with uncatchable exceptions, VS Test Explorer doesn't quite indicate that; instead of indicating that the test failed they appear as 'not run'. There's this small notification at the bottom of VS but it's not as noticeable as something like a proper pop-up, like the one on build failure.
 
![image](https://user-images.githubusercontent.com/42944058/89437394-7b258c00-d782-11ea-8170-fb7ee52f7b1a.png)

![image](https://user-images.githubusercontent.com/42944058/89437238-49142a00-d782-11ea-9682-281379dd6981.png)

If one decides to take a look at the test log, then you might be able to figure out it's an AVE thrown from the test method.

<details><summary>Example log: AVE</summary>

```
Log level is set to Informational (Default).
Test data store opened in 0.015 sec.
Build failed or it was cancelled.
---------- Starting test discovery for requested test run ----------
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.2+2d84eb3141 (64-bit .NET 5.0.0-rc.1.20401.1)
[xUnit.net 00:00:01.38]   Discovering: XUnitTestProject3
[xUnit.net 00:00:01.41]   Discovered:  XUnitTestProject3
========== Test discovery finished: 2 Tests found in 2.8 sec ==========
---------- Starting test run ----------
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.2+2d84eb3141 (64-bit .NET 5.0.0-rc.1.20401.1)
[xUnit.net 00:00:00.29]   Starting:    XUnitTestProject3
활성 테스트 실행이 중단되었습니다. 이유: 테스트 호스트 프로세스 작동이 중단됨 : Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at XUnitTestProject3.UnitTest1.ThrowsAve()
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.MethodBase.Invoke(System.Object, System.Object[])
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].CallTestMethod(System.Object)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<InvokeTestMethodAsync>b__1>d<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<InvokeTestMethodAsync>b__1>d<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<InvokeTestMethodAsync>b__1()
   at Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AggregateAsync>d__4 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AggregateAsync>d__4 ByRef)
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(System.Func`1<System.Threading.Tasks.Task>)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<InvokeTestMethodAsync>b__0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9, xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__9 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9, xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__9 ByRef)
   at Xunit.Sdk.ExceptionAggregator.RunAsync(System.Func`1<System.Threading.Tasks.Task>)
   at Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestMethodAsync>d__48<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestMethodAsync>d__48<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InvokeTestMethodAsync(System.Object)
   at Xunit.Sdk.XunitTestInvoker.InvokeTestMethodAsync(System.Object)
   at Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<RunAsync>b__47_0>d<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<RunAsync>b__47_0>d<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<RunAsync>b__47_0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__10`1<System.Decimal> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__10`1<System.Decimal> ByRef)
   at Xunit.Sdk.ExceptionAggregator.RunAsync[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Func`1<System.Threading.Tasks.Task`1<System.Decimal>>)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunAsync()
   at Xunit.Sdk.XunitTestRunner.InvokeTestMethodAsync(Xunit.Sdk.ExceptionAggregator)
   at Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestAsync>d__4 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestAsync>d__4 ByRef)
   at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(Xunit.Sdk.ExceptionAggregator)
   at Xunit.Sdk.TestRunner`1+<>c__DisplayClass43_0[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<RunAsync>b__0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`
========== Test run aborted: 0 Tests run in 3.2 sec (0 Passed, 0 Failed, 0 Skipped) ==========

```

</details>
<details><summary>Example log: Stack Overflow</summary>

```
---------- Starting test run ----------
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.2+2d84eb3141 (64-bit .NET 5.0.0-rc.1.20401.1)
[xUnit.net 00:00:00.29]   Starting:    XUnitTestProject3
활성 테스트 실행이 중단되었습니다. 이유: 테스트 호스트 프로세스 작동이 중단됨 : Stack overflow.
Repeat 32049 times:
--------------------------------
   at XUnitTestProject3.UnitTest1.StackOverflows()
--------------------------------
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.MethodBase.Invoke(System.Object, System.Object[])
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].CallTestMethod(System.Object)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<InvokeTestMethodAsync>b__1>d<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1+<<InvokeTestMethodAsync>b__1>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<InvokeTestMethodAsync>b__1>d<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<InvokeTestMethodAsync>b__1()
   at Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AggregateAsync>d__4 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.ExecutionTimer+<AggregateAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<AggregateAsync>d__4 ByRef)
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(System.Func`1<System.Threading.Tasks.Task>)
   at Xunit.Sdk.TestInvoker`1+<>c__DisplayClass48_1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<InvokeTestMethodAsync>b__0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9, xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__9 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__9, xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__9 ByRef)
   at Xunit.Sdk.ExceptionAggregator.RunAsync(System.Func`1<System.Threading.Tasks.Task>)
   at Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestMethodAsync>d__48<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestInvoker`1+<InvokeTestMethodAsync>d__48[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestMethodAsync>d__48<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InvokeTestMethodAsync(System.Object)
   at Xunit.Sdk.XunitTestInvoker.InvokeTestMethodAsync(System.Object)
   at Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<RunAsync>b__47_0>d<System.__Canon> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.TestInvoker`1+<<RunAsync>b__47_0>d[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<<RunAsync>b__47_0>d<System.__Canon> ByRef)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<RunAsync>b__47_0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__10`1<System.Decimal> ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<RunAsync>d__10`1<System.Decimal> ByRef)
   at Xunit.Sdk.ExceptionAggregator.RunAsync[[System.Decimal, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Func`1<System.Threading.Tasks.Task`1<System.Decimal>>)
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].RunAsync()
   at Xunit.Sdk.XunitTestRunner.InvokeTestMethodAsync(Xunit.Sdk.ExceptionAggregator)
   at Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestAsync>d__4 ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]](<InvokeTestAsync>d__4 ByRef)
   at Xunit.Sdk.XunitTestRunner.InvokeTestAsync(Xunit.Sdk.ExceptionAggregator)
   at Xunit.Sdk.TestRunner`1+<>c__DisplayClass43_0[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<RunAsync>b__0()
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.__Canon, System.Private.CoreL
========== Test run aborted: 0 Tests run in 19.8 sec (0 Passed, 0 Failed, 0 Skipped) ==========


```

</details>

Still, it's not immediately visible and I wish it could be improved by showing that it failed clearly, and telling which particular test it failed in. But then, I suppose it might be hard to diagnose this without being able to communicate with the test host process... allowing processes to catch those exceptions may make it possible to report enough data for the report before it `FailFast`s.

However, I agree with @​GrabYourPitchfork that it's just too dangerous to let people catch those unrecoverable exceptions and IMO even if we allowed it to be catchable it'll be too flaky of an environment to take any of these actions reliably. (e.g. it'd be incredibly hard to write code that can do those clean-ups in the same process, without allocating more memory on the heap / using more stack space / in possibly corrupted environment due to access violations).

**Comment by Gnbrkm41** [comment]
Filed https://developercommunity.visualstudio.com/idea/1137440/make-vs-test-explorer-report-clearer-messages-when.html for the VS test explorer issue.

I'm not exactly sure what happens with something like `dotnet test`, but since they use the same VSTest as the base I wouldn't be  surprised if it does the same with unexpected crashes of the test host...

**Comment by GrabYourPitchforks** [comment]
Off-the-wall brainstorming idea. Haven't thought about how feasible this might be.

Allow registering a global "stack overflow occurred" handler. If any thread encounters a stack overflow, the runtime will suspend that thread, and _on another thread_ invoke any registered `StackOverflowOccurredHandler`, passing as an argument the `System.Threading.Thread` corresponding to the thread that triggered the exception. They can inspect the thread object, capturing things like the call stack (how?) or logging information to disk. Then once the handlers have finished executing, the runtime terminates in the usual fashion.

**Comment by jkotas** [comment]
This may work if the stackoverflow happens while executing managed code itself. It does not work reliably for stackoverflows that happen inside the unmanaged runtime or unmanaged libraries. For example, the runtime or OS library may have lock taken (e.g. `malloc` lock) that will prevent everything else in the process from making process.

**Comment by Gnbrkm41** [comment]
Could be a viable option if we could make it super clear that it's for logging / clean-up situations only, then it probably would unblock the scenario for (managed) StackOverflowException I suppose. How about also providing something like a `StackTrace` object of that particular thread?

RE: deadlock - Just a thought, what if we impose some sort of time limit on it so it won't just stay deadlocked? probably not the best idea though.

I somewhat wished there's more of a general purpose mechanism for allowing unrecoverable exceptions to be logged though. This approach seems only applicable to StackOverflowException. But then again, there might be no silver bullet for all of them...

**Comment by DemiMarie** [comment]
What if we only allowed the exception to be caught if it was taken in managed code?  A stack overflow in native code is obviously fatal.

By using stack probes, we can ensure that when the `StackOverflowException` is thrown, we at least have enough stack space to throw the exception.```

**Comment by jinek** [comment]
> By using stack probes, we can ensure that when the `StackOverflowException` is thrown, we at least have enough stack space to throw the exception.

Why do we need to express fatal situations by Exceptions at all?

**Comment by VBAndCs** [comment]
How does vs.net discover SOF exception at debug time and highlight the line that caused it?
In test project, ruining the test crashes the test host but debugging it beaks at the line that caused the SOF.
Can the test host do something like this?.. may be run tests in another process and monitor it to report the failure?

**Comment by DemiMarie** [comment]
> > By using stack probes, we can ensure that when the `StackOverflowException` is thrown, we at least have enough stack space to throw the exception.
> 
> Why do we need to express fatal situations by Exceptions at all?

So that they can be caught.

**Comment by GSPP** [comment]
Another important scenario is a production error (possibly leading to an outage). Here, we need to resolve the outage as quickly as possible and then fix the issue. Making the exception catchable in an opt-in "emergency mode" seems like a great solution. This would make the usual catch-all error logging work again. Most apps have that.

**Comment by VBAndCs** [comment]
@CyrusNajmabadi 
There are some ideas here. What do you think.
And here is another one:
When a function calls itself for the first time, record the memory size avaliable in the stack, repeat the same for the second call, and hence get the minimal size needed per call, hence calculate the max allowed recursive calls (mc), and throw the Stack overflow at (95.0 * mc), so, there is still a room left to handle the exception. 
Of course the 95.0 ratio is an istimation and can be adjusted in pratical testing.

**Comment by jinek** [comment]
I have made related proposal which assumes possibility to catch SOF, ThreadAbort etc https://github.com/dotnet/runtime/issues/45256



> @CyrusNajmabadi There are some ideas here. 

We had similar discussion in another post. When we catch NullReferenceException there is no guarantee that another NullReferenceException will not be thrown from the catch block. Same is for every other exception. So here is the question, why do we want special behavior for SOF? Why do we want guarantee catch execution specifically for SOF exception? May be stack has been already unwound and has some space, may be no. What is so special about SOF? Or do we need space to evaluate catch blocks on the way?

**Comment by CyrusNajmabadi** [comment]
> Making the exception catchable in an opt-in "emergency mode" seems like a great solution.

I disagree.  Things are likely entirely corrupt and will likely do more damage of they proceed.  If you need to function in the face of catastrophic circumstances, then you can design your system to have a watchdog that can see when your main process dies and can take corrective action *external* to that process.

**Comment by jinek** [comment]
> > Making the exception catchable in an opt-in "emergency mode" seems like a great solution.
> 
> I disagree. Things are likely entirely corrupt and will likely do more damage of they proceed.

This is also true about any exception in general.

**Comment by CyrusNajmabadi** [comment]
> This is also true about any exception in general.

I don't believe this to be the case at all.

**Comment by jinek** [comment]
> > This is also true about any exception in general.
> 
> I don't believe this to be the case at all.

This is not like religion to believe in, it can be logically proven.
Program execution over undetermined state (which the exception signals about) comes to corrupted data in general case.
For example, like this 
```
class Program
    {
        static void Main(string[] args)
        {
            double sumOfSquaredRoot = 0;
            
            for (int i = 0; i < 3; i++)
            {
                try
                {
                    sumOfSquaredRoot += XSquaredCalculator.Calculate(i);
                }
                catch { /*logging here*/}
            }
            
            Console.WriteLine($"Sum of first 5 digits squared is {sumOfSquaredRoot}");
        }
    }

    public static class XSquaredCalculator
    {
        static XSquaredCalculator()
        {
            throw new InvalidOperationException("Random exception for data corruption");
        }

        public static double Calculate(double x)
        {
            return x * x;
        }
    }
```

**Comment by CyrusNajmabadi** [comment]
> which the exception signals about

I don't believe this to be the case at all.

You may be talking about a certain class of exceptions.  But it's definitely not inherent to exceptions (or how they're commonly utilized).

**Comment by jinek** [comment]
> > which the exception signals about
> 
> I don't believe this to be the case at all.
> 
> You may be talking about a certain class of exceptions. But it's definitely not inherent to exceptions (or how they're commonly utilized).

I made the case more obvious, with also the case of particular exception we recognize. It shows there is no difference between SOF and any other random exception

```
public void ChargeComission(decimal comission)
        {
            decimal balance = 0;

            try
            {
                balance = GetBalance();
            }
            catch (ZeroBalanceException)
            {
                // this is particular exception which assumes we should use 0 for the balance
            }
            catch (Exception e)
            {
                // It does not matter whether e is StackOverflowException or NullReferenceException
                // or IndexOutOfRangeException or ThreadAbortException, state is broken
                // and commenting out "throw" leads
                // to further data corruption
                throw;
            }

            balance -= comission;

            UpdateBalance(balance);
        }
```

**Comment by CyrusNajmabadi** [comment]
> I made the case more obvious, with also the case of particular exception we recognize. It shows there is no difference between SOF and any other random exception

There is absolutely a difference.  Take something as basic as `OperationCanceledException`.  It is not true that "Things are likely entirely corrupt and will likely do more damage of they proceed."

Indeed, if that were the case, programming with TPL would be nigh-on impossible.

**Comment by jinek** [comment]
> There is absolutely a difference. Take something as basic as `OperationCanceledException`. It is not true that "Things are likely entirely corrupt and will likely do more damage of they proceed."
> 
> Indeed, if that were the case, programming with TPL would be nigh-on impossible.

What I said is true about OperactionCancelledException as well. The issue is with "likely" or "not likely". In general case it gets corrupted even if it's not likely to.
TPL is a good example of things getting corrupted. Because of TPL, exceptions like ThreadAbortException can not work normally. While ThreadAbortException got special behavior at some point, customer defined exceptions, like CustomThreadAbortException or for instance BreakSchedulerLoopException still do not work.
More simple case is BackgroundService (which is based on TPL) completely looses exceptions.

**Comment by CyrusNajmabadi** [comment]
> What I said is true about OperactionCancelledException as well. The issue is with "likely" or "not likely".

This is incorrect. And, again, of this were the case tpl programming would just not be possible. 

In Roslyn itself there are likely hundreds (or more) cancellation exceptions per second. If those could corrupt state you'd be screwed immediately.

**Comment by CyrusNajmabadi** [comment]
> Don't catch an exception unless you can handle it and leave the application in a known state.

Correct. Which clearly indicates that it's possible to deal with exceptions safely when appropriate. Entire apis and programming models are designed around this. I'm not sure what else to tell you about this.

**Comment by jinek** [comment]
> Correct. Which clearly indicates that it's possible to deal with exceptions safely when appropriate. Entire apis and programming models are designed around this. I'm not sure what else to tell you about this.

It does not declare special rules for some exceptions. Catching SOF is same bad/good as anything else:
```
            foreach (var job in jobs)
            {
                try
                {
                    job.Execute();
                }
                catch (StackOverflowException)
                {//this catch can be ommited
                    Console.WriteLine($"SOF happened during execution of {job.Id}");
                }
                catch (Exception)
                {
                    Console.WriteLine($"Exception happened during execution of {job.Id}");
                }
            }

```

**Comment by CyrusNajmabadi** [comment]
> It does not declare special rules for some exceptions. 

@jinek I don't know what to tell you.  Catching exceptions is absolutely part and parcel of correct .net programming.  This is simply not true:

![image](https://user-images.githubusercontent.com/4564579/147858179-56f90475-1482-41cc-8bf2-a60e3a30b642.png)

Are there exceptions you should not catch?  Absolutely.  But your claim above is just incorrect, and there is ample code (written by us on runtime, roslyn, etc.) that demonstrates the appropriate and necessity of this.  As i've already pointed out, full frameworks are designed and documented with this being a core principal.

> Catching SOF is same bad/good as anything else:

This is not true.  And you keep repeating it without actually addressing the points that show it to not be true.  Catching SOF is not at all similar to catching "anything else".  For example, catching OperationCanceledException or TimeoutException are entirely different and not at all in teh same category as SOF.

**Comment by jinek** [comment]
> For example, catching OperationCanceledException or TimeoutException are entirely different and not at all in teh same category as SOF.

So what is the difference?

**Comment by jinek** [comment]
> > It does not declare special rules for some exceptions.
> 
> @jinek I don't know what to tell you. Catching exceptions is absolutely part and parcel of correct .net programming. This is simply not true:
> 
> ![image](https://user-images.githubusercontent.com/4564579/147858179-56f90475-1482-41cc-8bf2-a60e3a30b642.png)
> 
> Are there exceptions you should not catch? Absolutely. But your claim above is just incorrect, and there is ample code (written by us on runtime, roslyn, etc.) that demonstrates the appropriate and necessity of this. As i've already pointed out, full frameworks are designed and documented with this being a core principal.

There was a statement:
> Another important scenario is a production error (possibly leading to an outage). Here, we need to resolve the outage as quickly as possible and then fix the issue. Making the exception catchable in an opt-in "emergency mode" seems like a great solution. This would make the usual catch-all error logging work again. Most apps have that.

Which I understand as 

```
foreach (var job in jobs)
            {
                try
                {
                    job.Execute();
                }                
                catch (Exception e)
                {
                    _logger.LogException(e);
                }
            }
```

If jobs access shared resource, same time each job depends on previous jobs executions, then swallowed random exception from one of the job in general means that shared data will get corrupted by following jobs execution. I gave example which counts the sum. Sum value will be corrupted if we just catch and log the exception.

**Comment by CyrusNajmabadi** [comment]
@jinek as i mentioned about that:

![image](https://user-images.githubusercontent.com/4564579/147858646-326e377a-0ad5-4605-a61d-13f1947dd998.png)

**Comment by CyrusNajmabadi** [comment]
> So what is the difference?

One is not likely to leave the programming logic in a correct state, the others are.  That's the point :)

**Comment by jinek** [comment]
> > So what is the difference?
> 
> One is not likely to leave the programming logic in a correct state, the others are. That's the point :)

Could you please elaborate this point? why `catch(NullReferenceException)` likely to leave the correct state, but `catch(SOF)` is not?

**Comment by CyrusNajmabadi** [comment]
> Could you please elaborate this point? why catch(NullReferenceException) likely to leave the correct state,

I never said it would.  I think you're arguing against a misunderstanding about what was written.

**Comment by jinek** [comment]
> One is not likely to leave the programming logic in a correct state, the others are. That's the point :)
> 
> Things like timeout and cancellation are part of the _normal_ control flow around program logic. they are expected, accepted, and part and parcel of interacting with many APIs. Indeed, for these types of exceptions, it's the supported way for an external consumer get appropriate signals from an API they're working with (and there are other cases as well).
> 
> You can def argue that perhaps it would have been better if .net not conflated these sort of correct-control flow with 'exceptions'. But here we are.

If I call `plugin.ApplyFilter(picture)` I don't expect OperationCancelledException and it's not part of the flow, but I would expect SOF or OOM. I really don't understand how you draw the line what is normal and what is not. Again I don't understand what is so special about SOF. What exactly makes it so "likely" to corrupt the data?

**Comment by CyrusNajmabadi** [comment]
> I really don't understand how you draw the line what is normal and what is not. 

In general, what APIs (and general coding patterns define).  Plus, just some common sense :)   If TimeoutException corrupted state, literally no code with timeouts would be possible in .Net without rampant instability.  Same with cancellation.

**Comment by CyrusNajmabadi** [comment]
> What exactly makes it so "likely" to corrupt the data?

**Comment by CyrusNajmabadi** [comment]
>  but I would expect SOF 

Why would you expect SOF here?  What would you even do in that case?  Are you even in a hosted environment where you can catch it?  We haven't allowed it to even be caught in more than a decade in normal .Net environments.

**Comment by CyrusNajmabadi** [comment]
Note: SOF and OOM are really in that *very* special category as it's not only the case taht teh *user's* code will likely not be able to handle either properly, but it's also the case that the *runtime* will likely not be in a state where it can proceed.  Say the runtime needs to run some code.  How can it?  there's no stack available. Say it needs to jit some code?  How can it if there's no memory available?  etc. etc.

Note: this is why we distinguish OutOfMemoryException vs [InsufficientMemoryException](https://docs.microsoft.com/en-us/dotnet/api/system.insufficientmemoryexception?view=net-6.0#remarks).  There we state explicitly:

> Unlike OutOfMemoryException, InsufficientMemoryException is thrown before starting an operation, and thus does not imply state corruption. An application can catch this exception, throttle back its memory usage, and avoid actual out of memory conditions and their potential for corrupting program state.

**Comment by jinek** [comment]
> Note: SOF and OOM are really in that _very_ special category as it's not only the case taht teh _user's_ code will likely not be able to handle either properly, but it's also the case that the _runtime_ will likely not be in a state where it can proceed. 

Could you please give an example when OOM is not possible to handle correctly?

I have dealt with OOM all the way with .NET with no issues.
![image](https://user-images.githubusercontent.com/10516222/147859656-1d5f8eff-c61e-4aec-a482-74c0d30c6f58.png)

Again, why do you say that user's code will likely not be able to handle them? I'm trying to figure out what so special happens with SOF or OOM?

**Comment by jinek** [comment]
> > but I would expect SOF
> 
> Why would you expect SOF here? What would you even do in that case? Are you even in a hosted environment where you can catch it? We haven't allowed it to even be caught in more than a decade in normal .Net environments.

For example, I'm calling a plugin and I know that plugin may have no enough stack to execute the operation. Once we made data to be processed by plugin isolated, we can just warn the user and allow other operations to be processed. The user can reduce the image size and repeat the operation.
Regarding hosted environment, that is the point, why not to make it in normal .NET. I've described the situation which assumes normal control flow dealing with SOF or OOM

**Comment by CyrusNajmabadi** [comment]
> Could you please give an example when OOM is not possible to handle correctly?

Imagine the runtime it out of memory.  And you call Console.WriteLine.  And `WriteLine` isn't even jitted yet.  The runtime cannot allocate memory to even do whatever you want in response to the problem.

> I have dealt with OOM all the way with .NET with no issues.

You are conflating two scenarios.  One is that the runtime actually has memory to proceed after the errant problem, the other is where it does not.  You cannot distinguish that unless using the very special procedure outlined in the docs i listed for InsufficientMemoryException.

**Comment by CyrusNajmabadi** [comment]
>  we can just warn the user 

How?  You may literally be out of stack.

> and allow other operations to be processed

What operations do you know can be processed?  Teh runtime may literally lack *any* resources to do any further processing.

> I've described the situation which assumes normal control flow dealing with SOF or OOM

No.  You've defined cases which *might* work, or which may go horribly wrong.

**Comment by CyrusNajmabadi** [comment]
> why not to make it in normal .NET. 

This was what was tried.  It turns out that in *in general* (not in specialized cases like you are identifying) this cannot be done.  Again, what do you do if you're literally out of memory and you cannot even JIT the recovery code to execute?  What do you do if you're literally out of stack and you need to allocate stack space?

Writing code that could actually behave correctly here *for all the myriad complex cases* turned out to be far too complex for the ecosystem to handle (and was indeed super hard even for MS as a first party to do effectively).  Please look into things like CERs (ConstrainedExecutionRegions) and how this was attempted but ultimately was just not feasible because it's so complex and challenging to do right for the general case.

Instead, we have *specialized* cases which *may* work in some very narrowly defined places.  They may even work well enough for certain domains (which can impose additional restrictions on the environment).  But they do not work out in the general case for the simple reasons i gave above. 

--

Again, just ask yourself: how can i proceed if the runtime literally does not have the resources it needs to proceed?

**Comment by jinek** [comment]
> Imagine the runtime it out of memory. And you call Console.WriteLine. And `WriteLine` isn't even jitted yet. The runtime cannot allocate memory to even do whatever you want in response to the problem.

Same about any exception! Even if you are catching `OperationCancelledException` WriteLine may be have no enough memory to allocate the char array. For any exception there is no guarantee the catch block to be executed. It is exactly same for OOM, is not it?

> You are conflating two scenarios. One is that the runtime actually has memory to proceed after the errant problem, the other is where it does not. You cannot distinguish that unless using the very special procedure outlined in the docs i listed for InsufficientMemoryException.

I don't see any reason why no to unwind the stack to some catch block and not to allow allocate some memory by executing GC. The situation when runtime can not JIT `WriteLine` is possible when executing catch block for any exception, agian. No difference if it is `OperationCancelledException` or OOM or anything else.

**Comment by CyrusNajmabadi** [comment]
> Even if you are catching OperationCancelledException WriteLine may be have no enough memory to allocate the char array

That's not true.  There's no relation between OCE and not having enough memory.

> For any exception there is no guarantee the catch block to be executed. It is exactly same for OOM, is not it?

That is not the case.

> The situation when runtime can not JIT WriteLine is possible when executing catch block for any exception, agian. 

It is not.  When encountering OOM, the runtime is already starved on resources (same as SOF).  This is not hte case with something like OCE.  There's no starvation.  It's an explicit part of the contracts around TPL on how OCE is used.

> No difference if it is OperationCancelledException or OOM or anything else.

I genuinely don't know what to say at this point.  I've linked our docs.  I've discussed the history here.  I've explained the distinction between exceptions thrown when teh runtime is out of resources (vs when it's not).  I've explained how there are absolutely exceptions that are *designed* into APIs as part of the signalling and control flow mechanisms.

You seem to be operating under your own conceptions of how this works.  And no amount of actual docs or explanations seems to be satisfactory for you.  I don't know what else i can add.

**Comment by CyrusNajmabadi** [comment]
I have to ask @jinek are you familiar with the design of the TPL?

**Comment by jinek** [comment]
> How? You may literally be out of stack.

With any exception we may be out of stack. Same time I can propose to unwind reserved space and allocate "user warning" frame instead.

> What operations do you know can be processed? Teh runtime may literally lack _any_ resources to do any further processing.

You can not ever know any operations to be processed. Even if you catch `PluginException` (which is part of API) it's still possible that all other operations will fail. May be runtime has no enough resources for some operations, but has for others.

**Comment by CyrusNajmabadi** [comment]
> With any exception we may be out of stack. Same time I can propose to unwind reserved space and allocate "user warning" frame instead.

This was the idea behind ConstrainedExecutionRegions.  However, they are *extremely* difficult to get right, and now are effectively non-grata.

> You can not ever know any operations to be processed.

Again, this is not the case.  And if it were it woudl not be possible to effectively do async processing.  

> it's still possible that all other operations will fail. 

You are conflating issues.  It's like saying "you can't continue if something returns you a value, because the very next step may fail".  Those are two separate steps.  

With things like OCE/Timeout/etc. you absolutely can proceed *and that's part of the design of .Net*.  I simply do not know how to state it any more plainly.  It would literally not be possible to handle timeout in .net if this were not the case. 

It would not be possible to do cancellation in .net if this were not the case.  

And yet, both of tehse (and many other cases) are amply doc'ed as being 100% supported and *expected* as the mechanism for how components can interact.  That is what the docs show that i've linked you to, and you insisting otherwise serves no purpose except to indicate that you are not willing ot actually read or engage with the things we've actually documented and published.

**Comment by CyrusNajmabadi** [comment]
> May be runtime has no enough resources for some operations, but has for others.

That is not relevant to the question of: can you handle *this* issue for things like OCE/Timeout.

For things like OOM/SOF.  It *is* very relevant, and it's precisely why we don't let you catch the latter and why we doc that catching the former is effectively not stable.  If you're going to ignore the docs on this, i genuinely just do not know what i can tell you.

**Comment by CyrusNajmabadi** [comment]
@jinek i recommend we take this to discord.gg/csharp.  This conversation has effectively gone in circles and seems to be doing nowhere.  It is clogging this issue explaining the current way .net works (and is not likely to change in).  If you'd like to discuss this more, we can take it there.

**Comment by jinek** [comment]
> > Even if you are catching OperationCancelledException WriteLine may be have no enough memory to allocate the char array
> 
> That's not true. There's no relation between OCE and running out of memory (or other resources).

I don't know why you think you can expect any given catch block not to throw OOM.
But as a proof here you can try yourself
![image](https://user-images.githubusercontent.com/10516222/147861841-2b953fd7-939d-4716-b8b8-b248ce4b9e4f.png)

**Comment by CyrusNajmabadi** [comment]
> I don't know why you think you can expect any given catch block not to throw OOM.

I never said that.  Furthermore your example proves my point.  The program is a totally fine state after the OCE (which is the normal expected case *as per the TPL* docs).  I don't know why you keep glossing over this. We have doc'ed this. We have ample experience trying different routes and seeing how it has/hasn't worked out in practice.  At this point i'm just not certain if you're even considering or looking at our docs here.  

Like i said, this isn't going anywhere.  If you want to keep up with this tangent, take it discord.  Further talking aobut it will be viewed as off-topic.

**Comment by jinek** [comment]
> Like i said, this isn't going anywhere. If you want to keep up with this tangent, take it discord. Further talking about things that directly contradict our stated docs here will be viewed as off-topic.

I think we can finish at this point, we stuck on simple false = not true.
We can switch to one-to-one if you think makes sense. I don't see a point to discuss it with .NET people exactly because of how framework is already implemented. TPL assumes it's possible to recover by re-throwing the exception on caller thread, while exception filters just swallow the exception completely. Both will get broken if we make SOF possible to catch (similar to OOM as it is now).

**Comment by CyrusNajmabadi** [comment]
>  Both will get broken if we make SOF possible to catch

I've never argued that SOF should be possible to catch.  Please read my posts.  All i said was:

![image](https://user-images.githubusercontent.com/4564579/147863423-f0665d29-252a-4a92-86c9-deb36d3cc38d.png)

I've been clear, this is in the category of errors that are likely just not at all possible to catch safely.  This is based on our actual experience in this realm.  I have maintained this, so I think you're arguing against something different with me.

--

> We can switch to one-to-one if you think makes sense.

Yes please.

**Comment by MichalStrehovsky** [comment]
> I don't see any reason why no to unwind the stack to some catch block

As an aside, when managed exception handling runs, the stack is not unwound yet. In particular, the runtime would need to be able to run exception filters (`catch (X) when (Y)`) before the stack starts getting unwound to run catch blocks. C# kind of forces you to do a method call in the `when` clause if one wants to do anything interesting (and also many things that don't look like calls are actually calls, like `as` and `is` checks), but since we don't have stack, it would probably throw stack overflow again.

Catch blocks associated with filters that throw get silently skipped.

The program is in an interesting state at that point.

**Comment by DemiMarie** [comment]
The standard solution to this in unmanaged code is an alternate signal stack.  Could the code be run on an alternate stack?  Furthermore, test frameworks might well be fine with skipping exception filters.

**Comment by GSPP** [comment]
I have made a list collecting ideas for how to improve the situation.

- https://github.com/dotnet/runtime/issues/4113#issuecomment-1000837221

If I missed something please let me know and I'll add it.

**Comment by jinek** [comment]
> Catch blocks associated with filters that throw get silently skipped.
> 
> The program is in an interesting state at that point.

That was my point: we need to think to fix this first, because possibility to catch SOF will add randomness to already random behavior (OutOfMemoryException + "catch when")
